src/util/sharing_node.h

```diff
@@ -12,339 +12,455 @@ Author: Daniel Poetzl
 #ifndef CPROVER_UTIL_SHARING_NODE_H
 #define CPROVER_UTIL_SHARING_NODE_H
 
-#include <list>
+//#define DEBUG
+
+#ifdef DEBUG
+#include <iostream>
+#endif
+
+#include <forward_list>
+#include <type_traits>
+
+#ifndef SN_SMALL_MAP
+#define SN_SMALL_MAP 1
+#endif
+
+#ifndef SN_SHARE_KEYS
+#define SN_SHARE_KEYS 0
+#endif
+
+#if SN_SMALL_MAP == 1
+#include "small_map.h"
+#else
 #include <map>
-#include <memory>
+#endif
 
 #include "invariant.h"
+#include "make_unique.h"
+#include "small_shared_ptr.h"
+#include "small_shared_two_way_ptr.h"
+
+//#define SN_INTERNAL_CHECKS
+
+#ifdef SN_INTERNAL_CHECKS
+#define SN_ASSERT(b) INVARIANT(b, "Sharing node internal invariant")
+#define SN_ASSERT_USE(v, b) SN_ASSERT(b)
+#else
+#define SN_ASSERT(b)
+#define SN_ASSERT_USE(v, b) v = v;
+#endif
+
+#define SN_TYPE_PAR_DECL                                                       \
+  template <typename keyT,                                                     \
+            typename valueT,                                                   \
+            typename equalT = std::equal_to<keyT>>
+
+#define SN_TYPE_PAR_DEF                                                        \
+  template <typename keyT, typename valueT, typename equalT>
+
+#define SN_TYPE_ARGS keyT, valueT, equalT
+
+#define SN_PTR_TYPE_ARGS d_internalt<SN_TYPE_ARGS>, d_containert<SN_TYPE_ARGS>
 
-#define _sn_assert(b) INVARIANT(b, "Sharing-node internal invariant")
-//#define _sn_assert(b)
+#define SN_PTR_TYPE_ARG d_leaft<SN_TYPE_ARGS>
 
 template <class T>
 const T *as_const(T *t)
 {
   return t;
 }
 
-template <
-  class keyT,
-  class valueT,
-  class predT=std::equal_to<keyT>,
-  bool no_sharing=false>
-class sharing_nodet
-{
-public:
-  friend void sharing_node_test();
+// Inner nodes (internal nodes or container nodes)
 
-  typedef keyT key_type;
-  typedef valueT mapped_type;
+typedef small_shared_two_way_pointeet<unsigned> d_baset;
 
-  typedef predT key_equal;
+SN_TYPE_PAR_DECL class sharing_node_innert;
 
-  typedef sharing_nodet<key_type, mapped_type, key_equal> self_type;
+SN_TYPE_PAR_DECL class d_internalt : public d_baset
```

Comment:
I think there is a question regarding the naming of this and similar classes. What does the `d_` prefix symbolise? Is it just `data`? Or ?

Reply:
I saw now, that these are supposed to mimic the naming scheme of `irept`, with `d` symbolising `data` or `data type` or something along this lines. Understood, but I think this should be documented somewhere more prominently than an easily forgotten comment in a PR.